### PR TITLE
state in viewmodel

### PR DESCRIPTION
**Pull request by lemppo**

### CODE REVIEW

File: src/main/java/ru/touchin/roboswag/components/navigation/fragments/ViewControllerFragment.java [hunk 1]
```diff
@@ -60,14 +66,32 @@
...
+        extends ViewFragment {
 
     private static final String VIEW_CONTROLLER_STATE_EXTRA = "VIEW_CONTROLLER_STATE_EXTRA";
+    private static final String VIEW_CONTROLLER_CLASS_EXTRA = "VIEW_CONTROLLER_CLASS_EXTRA";
+    private static final String VIEW_MODEL_CLASS_EXTRA = "VIEW_MODEL_CLASS_EXTRA";
 
     private static boolean inDebugMode;
     private static long acceptableUiCalculationTime = 100;
 
+    private Class<? extends ViewController<TActivity, ? extends ViewControllerFragment<TState, TActivity>>> viewControllerClass;
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/navigation/fragments/ViewControllerFragment.java]: Move all nonstatic fields after static methods

File: src/main/java/ru/touchin/roboswag/components/navigation/fragments/ViewControllerFragment.java [hunk 2]
```diff
@@ -60,14 +66,32 @@
...
+    private Class<? extends StateViewModel> viewModelClass;
+
+    @NonNull
+    private final BehaviorSubject<Optional<ViewControllerActivity>> activitySubject = BehaviorSubject.create();
+    @NonNull
+    private final BehaviorSubject<NullablePair<PlaceholderView, Bundle>> viewSubject = BehaviorSubject.create();
+    @Nullable
+
+    private ViewController viewController;
+    private StateViewModel<TState> viewModel;
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/navigation/fragments/ViewControllerFragment.java]: local variable?

File: src/main/java/ru/touchin/roboswag/components/navigation/fragments/ViewControllerFragment.java [hunk 3]
```diff
@@ -84,6 +108,50 @@ public static void setAcceptableUiCalculationTime(final long acceptableUiCalcula
...
+    public static Bundle createState(@NonNull final Class<? extends ViewController> viewControllerClass,
+                                     @NonNull final Class<? extends ViewModel> viewModelClass,
+                                     @NonNull final AbstractState state) {
+        final Bundle result = createState(state);
+        result.putSerializable(VIEW_CONTROLLER_CLASS_EXTRA, viewControllerClass);
+        result.putSerializable(VIEW_MODEL_CLASS_EXTRA, viewModelClass);
+        return result;
+    }
+
+    public StateViewModel getViewModel() {
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/navigation/fragments/ViewControllerFragment.java]: Unnecessary

File: src/main/java/ru/touchin/roboswag/components/utils/lifecycle/BaseDestroyable.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                    Observable.empty<T>()
+                }
+                .subscribe()
+    }
+
+    private fun getActionThrowableForAssertion(codePoint: String, method: String): Consumer<Throwable> {
+        return Consumer { t -> Lc.assertion(ShouldNotHappenException("Unexpected error on $method at $codePoint", t)) }
+    }
+
+    companion object {
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/utils/lifecycle/BaseDestroyable.kt]: to start of class

File: src/main/java/ru/touchin/roboswag/components/utils/lifecycle/BaseStopable.kt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                               onSuccessAction: Consumer<T>,
+                               onErrorAction: Consumer<Throwable>): Disposable {
+        return until(maybe.toObservable(), isStartedSubject.map { started -> !started }, onSuccessAction, onErrorAction, Functions.EMPTY_ACTION)
+    }
+
+    private fun getActionThrowableForAssertion(codePoint: String, method: String): Consumer<Throwable> {
+        return Consumer { t -> Lc.assertion(ShouldNotHappenException("Unexpected error on $method at $codePoint", t)) }
+    }
+
+    companion object {
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/utils/lifecycle/BaseStopable.kt]: to start of class

File: src/main/java/ru/touchin/roboswag/components/viewmodel/StateViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import ru.touchin.roboswag.components.utils.lifecycle.Destroyable
+import ru.touchin.roboswag.components.navigation.AbstractState
+import ru.touchin.roboswag.core.log.Lc
+import ru.touchin.roboswag.core.utils.ShouldNotHappenException
+
+open class StateViewModel<TState : AbstractState>(application: Application) : AndroidViewModel(application), Destroyable {
+
+    private val baseDestroyable = BaseDestroyable()
+
+    var state: TState? = null
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/viewmodel/StateViewModel.kt]: попробуй lateinit var, в котлине начиная с 1.2 можно проверить ::state.isInitialized

File: src/main/java/ru/touchin/roboswag/components/viewmodel/StateViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import io.reactivex.functions.Consumer
+import ru.touchin.roboswag.components.utils.lifecycle.BaseDestroyable
+import ru.touchin.roboswag.components.utils.lifecycle.Destroyable
+import ru.touchin.roboswag.components.navigation.AbstractState
+import ru.touchin.roboswag.core.log.Lc
+import ru.touchin.roboswag.core.utils.ShouldNotHappenException
+
+open class StateViewModel<TState : AbstractState>(application: Application) : AndroidViewModel(application), Destroyable {
+
+    private val baseDestroyable = BaseDestroyable()
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/viewmodel/StateViewModel.kt]: to constructor

File: src/main/java/ru/touchin/roboswag/components/viewmodel/StateViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+import io.reactivex.disposables.Disposable
+import io.reactivex.functions.Action
+import io.reactivex.functions.Consumer
+import ru.touchin.roboswag.components.utils.lifecycle.BaseDestroyable
+import ru.touchin.roboswag.components.utils.lifecycle.Destroyable
+import ru.touchin.roboswag.components.navigation.AbstractState
+import ru.touchin.roboswag.core.log.Lc
+import ru.touchin.roboswag.core.utils.ShouldNotHappenException
+
+open class StateViewModel<TState : AbstractState>(application: Application) : AndroidViewModel(application), Destroyable {
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/viewmodel/StateViewModel.kt]: Destroyable by baseDestroyable

File: src/main/java/ru/touchin/roboswag/components/viewmodel/StateViewModel.kt (hunk 6 above)
Comment (lemppo) [reply src/main/java/ru/touchin/roboswag/components/viewmodel/StateViewModel.kt]: lateinit is not allowed for properties with custom getter and setter

File: src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    : RecyclerView.ViewHolder(itemView), Stopable by baseStopable {
+
+    /**
+     * Look for a child view with the given id.  If this view has the given id, return this view.
+     *
+     * @param id The id to search for;
+     * @return The view that has the given id in the hierarchy.
+     */
+    fun <T : View> findViewById(@IdRes id: Int): T? =
+        itemView.findViewById<View>(id) as T?
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt]: itemView.findViewById(id)

File: src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                ?: throw ShouldNotHappenException("No view for id=" + itemView.resources.getResourceName(id))
+
+    /**
+     * Return the string value associated with a particular resource ID.  It
+     * will be stripped of any styled text information.
+     *
+     * @param resId The resource id to search for data;
+     * @return String The string data associated with the resource.
+     */
+    fun getString(@StringRes resId: Int): String =
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt]: in single line

File: src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Return the string value associated with a particular resource ID.  It
+     * will be stripped of any styled text information.
+     *
+     * @param resId      The resource id to search for data;
+     * @param formatArgs The format arguments that will be used for substitution.
+     * @return String The string data associated with the resource.
+     */
+    fun getString(@StringRes resId: Int, vararg formatArgs: Any): String =
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt]: in single line

File: src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Return the color value associated with a particular resource ID.
+     * Starting in [android.os.Build.VERSION_CODES.M], the returned
+     * color will be styled for the specified Context's theme.
+     *
+     * @param resId The resource id to search for data;
+     * @return int A single color value in the form 0xAARRGGBB.
+     */
+    @ColorInt
+    fun getColor(@ColorRes resId: Int): Int =
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt]: in single line

File: src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Returns a drawable object associated with a particular resource ID.
+     * Starting in [android.os.Build.VERSION_CODES.LOLLIPOP], the
+     * returned drawable will be styled for the specified Context's theme.
+     *
+     * @param resId The resource id to search for data;
+     * @return Drawable An object that can be used to draw this resource.
+     */
+    fun getDrawable(@DrawableRes resId: Int): Drawable? =
```
Comment (Denis55ka) [inline src/main/java/ru/touchin/roboswag/components/adapters/BindableViewHolder.kt]: in single line

**Pull request closed without merging**